django/db/migrations/state.py

```diff
@@ -94,16 +94,30 @@ def __init__(self, models=None, real_apps=None):
         self.real_apps = real_apps or []
         self.is_delayed = False
         # {remote_model_key: {model_key: [(field_name, field)]}}
-        self.relations = None
+        self.relations = defaultdict(partial(defaultdict, list))
 
     def add_model(self, model_state):
         app_label, model_name = model_state.app_label, model_state.name_lower
         self.models[(app_label, model_name)] = model_state
+        if 'all_relations' in self.__dict__:
+            concretes, _ = self._get_concrete_models_mapping_and_proxy_models()
+            self.populate_relation_from_model_state(model_state, (app_label, model_name), concretes)
```

Comment:
`(app_label, model_name)` is also used to get a model state, I'd cache it in a local variable, e.g.:
```python
model_key = model_state.app_label, model_state.name_lower
self.models[model_key] = model_state
if self._relations is not None:
    concretes, _ = self._get_concrete_models_mapping_and_proxy_models()
    self.populate_relation_from_model_state(model_state, model_key, concretes)
if 'apps' in self.__dict__:  # hasattr would cache the property
    self.reload_model(*model_key)
```